pkg/ebpf/c/tracee.bpf.c

```diff
@@ -4295,202 +4416,122 @@ int BPF_KPROBE(cgroup_bpf_run_filter_skb)
-    struct entry *entry = bpf_map_lookup_elem(&entrymap, &host_tid);
-    if (!entry) // no entry == no tracing
-        return 0;
-
-    // pick args from entry point's entry
-    // struct sock *sk = (void *) entry->args[0];
-    struct sk_buff *skb = (void *) entry->args[1];
-
-    // cleanup entrymap
-    bpf_map_delete_elem(&entrymap, &host_tid);
-
-    // use skb timestamp as the key for cgroup/skb
-    u64 skbts = BPF_READ(skb, tstamp);
-
-    // delete netctx after cgroup ebpf program runs
-    bpf_map_delete_elem(&cgrpctxmap, &skbts);
-
-    return 0;
-}
+    // read IP/IPv6 headers
 
-//
-// Type definitions and prototypes for protocol parsing
-//
-
-// NOTE: proto header structs need full type in vmlinux.h (for correct skb copy)
-
-typedef union iphdrs_t {
-    struct iphdr iphdr;
-    struct ipv6hdr ipv6hdr;
-} iphdrs;
+    void *data_ptr = NULL;
+    u16 mac_len = BPF_READ(skb, mac_len);
+    if (!mac_len) {
+        data_ptr = BPF_READ(skb, data); // no L2 header present in skb
+    } else {
+        data_ptr = BPF_READ(skb, head);
+        u16 nethead = BPF_READ(skb, network_header);
+        data_ptr += nethead;
+    }
+    bpf_core_read(nethdrs, l3_size, data_ptr);
 
-typedef union protohdrs_t {
-    struct tcphdr tcphdr;
-    struct udphdr udphdr;
-    struct icmphdr icmphdr;
-    struct icmp6hdr icmp6hdr;
-    union {
-        u8 tcp_extra[40]; // data offset might set it up to 60 bytes
-    };
-} protohdrs;
+    // prepare the indexer with IP/IPv6 headers
 
-typedef struct nethdrs_t {
-    iphdrs iphdrs;
-    protohdrs protohdrs;
-} nethdrs;
+    u8 proto = 0;
 
-// clang-format off
+    indexer_t indexer = {0};
+    indexer.ts = BPF_READ(skb, tstamp);
 
-#define CGROUP_SKB_HANDLE_FUNCTION(name)                                       \
-static __always_inline u32 cgroup_skb_handle_##name(                           \
-    struct __sk_buff *ctx,                                                     \
-    net_event_context_t *neteventctx,                                          \
-    nethdrs *nethdrs                                                           \
-)
+    switch (family) {
+        case PF_INET:
+            if (nethdrs->iphdrs.iphdr.version != 4) // IPv4
+                return 1;
 
-CGROUP_SKB_HANDLE_FUNCTION(family);
-CGROUP_SKB_HANDLE_FUNCTION(proto);
-CGROUP_SKB_HANDLE_FUNCTION(proto_tcp);
-CGROUP_SKB_HANDLE_FUNCTION(proto_tcp_dns);
-CGROUP_SKB_HANDLE_FUNCTION(proto_tcp_http);
-CGROUP_SKB_HANDLE_FUNCTION(proto_udp);
-CGROUP_SKB_HANDLE_FUNCTION(proto_udp_dns);
-CGROUP_SKB_HANDLE_FUNCTION(proto_icmp);
-CGROUP_SKB_HANDLE_FUNCTION(proto_icmpv6);
+           if (nethdrs->iphdrs.iphdr.ihl > 5) { // re-read IP header if needed
+                l3_size -= get_type_size(struct iphdr);
+                l3_size += nethdrs->iphdrs.iphdr.ihl * 4;
+                bpf_core_read(nethdrs, l3_size, data_ptr);
+            }
 
-#define CGROUP_SKB_HANDLE(name) cgroup_skb_handle_##name(ctx, neteventctx, nethdrs);
```

Comment:
This is just a positional change. I'll eventually spinoff the network code into its own file anyway.